hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvokeOnPartitions.java

```diff
@@ -16,137 +16,212 @@
 import static com.hazelcast.util.CollectionUtil.toIntArray;
 import static com.hazelcast.util.MapUtil.createHashMap;
 
 /**
  * Executes an operation on a set of partitions.
- *
- * // TODO delete this class in 3.12, replace with {@link InvokeOnPartitionsAsync}
- * which has the same functionality, but in an async implementation. Allows
- * for async version of all-partitions operations, such as putAll, clear...
  */
 final class InvokeOnPartitions {
 
     private static final int TRY_COUNT = 10;
     private static final int TRY_PAUSE_MILLIS = 300;
 
-    private OperationServiceImpl operationService;
+    private static final Object NULL_RESULT = new Object() {
+        @Override
+        public String toString() {
+            return "NULL_RESULT";
+        }
+    };
+
+    private final OperationServiceImpl operationService;
     private final String serviceName;
     private final OperationFactory operationFactory;
     private final Map<Address, List<Integer>> memberPartitions;
-    private final Map<Address, Future> futures;
-    private final Map<Integer, Object> partitionResults;
+    private final AtomicReferenceArray<Object> partitionResults;
+    private final AtomicInteger latch;
+    private volatile ExecutionCallback<Map<Integer, Object>> callback;
+    private final SimpleCompletableFuture future;
 
     InvokeOnPartitions(OperationServiceImpl operationService, String serviceName, OperationFactory operationFactory,
                        Map<Address, List<Integer>> memberPartitions) {
         this.operationService = operationService;
         this.serviceName = serviceName;
         this.operationFactory = operationFactory;
         this.memberPartitions = memberPartitions;
-        this.futures = createHashMap(memberPartitions.size());
         int partitionCount = operationService.nodeEngine.getPartitionService().getPartitionCount();
-        this.partitionResults = createHashMap(partitionCount);
+        // this is the total number of partitions for which we actually have operation
+        int actualPartitionCount = 0;
+        for (List<Integer> mp : memberPartitions.values()) {
+            actualPartitionCount += mp.size();
+        }
+        this.partitionResults = new AtomicReferenceArray<Object>(partitionCount);
+        this.latch = new AtomicInteger(actualPartitionCount);
+        this.future = new SimpleCompletableFuture(operationService.nodeEngine);
     }
 
     /**
      * Executes all the operations on the partitions.
      */
     <T> Map<Integer, T> invoke() throws Exception {
-        ensureNotCallingFromPartitionOperationThread();
+        return this.<T>invokeAsync(null).get();
+    }
 
+    /**
+     * Executes all the operations on the partitions.
+     */
+    @SuppressWarnings("unchecked")
+    <T> ICompletableFuture<Map<Integer, T>> invokeAsync(ExecutionCallback<Map<Integer, T>> callback) {
```

Comment:
Do we really need this additional `callback` parameter? Looks like the same behaviour can be handled by simple `andThen` on the caller side. Additionally, we will avoid ambiguity between the order of this callback and extra `andThen`s the caller might have installed (now the callback is executed after all `andThen`s).

Reply:
I've got similar comment several times. My response is: some async operations use callback, others use the returned future. Of course, they are replacable by each other. But why convert the api?

For example, `IMap.submitToKey` is available in both flavors, async version of which I have been creating. I don't know why is this. Tell me which one I should keep and I'll do it.

Reply:
The problem is that one may use `invokeAsync` in those two ways simultaneously and they are slightly conflicting, `submitToKey` is either accepting a callback or returning a future. Not a big deal for me anyway :)